### PR TITLE
feat(productcatalog): enforce rate card feature references

### PR DESCRIPTION
**Pull request by chrisgacsal**
## What

Enforce feature-reference integrity for plan and add-on rate cards so `feature_id` and `feature_key` are either both populated or both absent.

## Why

The existing backfill repairs incomplete references created by the former resolution bug. Database constraints are needed to prevent invalid rate cards from being persisted again.

## How

- Add Ent-side non-empty validation for both fields.
- Add PostgreSQL `CHECK` constraints for plan and add-on rate cards.
- Generate the corresponding Atlas migration.
- Add database-backed tests covering valid, incomplete, empty, pre-existing, and rollback scenarios.
- Update the migration guide with the enforced constraint contract.

JIRA: OM-464

## Summary by CodeRabbit

* **New Features**
  * Added validation ensuring feature references are complete and non-empty when provided.
  * Added database safeguards for plan and add-on rate-card feature references.

* **Bug Fixes**
  * Prevented incomplete, empty, or mismatched feature references from being saved.

* **Tests**
  * Added coverage for valid references, invalid data, migration failures, and rollback behavior.

* **Documentation**
  * Updated migration guidance to reflect the implemented validation and database safeguards.

<h3>Greptile Summary</h3>

The PR enforces complete, non-empty feature references for plan and add-on rate cards at both the Ent and PostgreSQL layers.
- Adds paired feature-reference checks to the authoritative Ent schemas and generated clients.
- Adds a checksummed migration and rollback for both rate-card tables.
- Adds database-backed coverage for valid, invalid, pre-existing, and rolled-back states.
- Documents the migration’s rollout prerequisites and enforced persistence contract.

<h3>Confidence Score: 5/5</h3>

The PR appears safe to merge.

No blocking failure remains.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| openmeter/ent/schema/ratecard.go | Adds non-empty validation for the shared optional feature key. |
| openmeter/ent/schema/productcatalog.go | Adds non-empty plan-rate-card feature IDs and the paired database constraint. |
| openmeter/ent/schema/addon.go | Adds non-empty add-on-rate-card feature IDs and the paired database constraint. |
| tools/migrate/migrations/20260818090123_rate_card_feature_reference_constraints.up.sql | Installs validated checks that accept only absent or fully populated feature references. |
| tools/migrate/migrations/20260818090123_rate_card_feature_reference_constraints.down.sql | Removes both feature-reference checks during rollback. |
| tools/migrate/rate_card_feature_reference_constraints_test.go | Exercises migration blocking, accepted reference shapes, rejected invalid shapes, and rollback behavior. |
| docs/migration-guides/2026-08-06-feature-reference-integrity.md | Documents the implemented constraints and the operational prerequisites for enabling them. |

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    W[Rate-card writer] --> E[Ent non-empty validation]
    E -->|Empty value| R[Reject mutation]
    E -->|Values absent or non-empty| D[(PostgreSQL)]
    D --> C{Paired-reference CHECK}
    C -->|Both absent| A[Accept row]
    C -->|Both non-empty| A
    C -->|Incomplete reference| X[Reject write]
    M[Migration] --> C
    L[Existing incomplete row] -->|Blocks validation| M
```
</details>

<sub>Reviews (2): Last reviewed commit: ["test(migrate): cover empty rate card fea..."](https://github.com/openmeterio/openmeter/commit/b5e43d2c32a3f59b289c7e8527854c57e80005a0) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=54378522)</sub>

**Context used:**

- Knowledge Base — [Data layer: ent schema, migrations, and generated client](https://app.greptile.com/openmeter/-/custom-context/knowledge-base/openmeterio/openmeter/-/docs/data-layer.md)

### CODE REVIEW

File: tools/migrate/migrations/20260818090123_rate_card_feature_reference_constraints.up.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+-- modify "addon_rate_cards" table
+ALTER TABLE "addon_rate_cards" ADD CONSTRAINT "addon_rate_card_feature_reference" CHECK (((feature_key IS NULL) AND (feature_id IS NULL)) OR ((feature_key IS NOT NULL) AND ((feature_key)::text <> ''::text) AND (feature_id IS NOT NULL) AND (feature_id <> ''::bpchar)));
+-- modify "plan_rate_cards" table
+ALTER TABLE "plan_rate_cards" ADD CONSTRAINT "plan_rate_card_feature_reference" CHECK (((feature_key IS NULL) AND (feature_id IS NULL)) OR ((feature_key IS NOT NULL) AND ((feature_key)::text <> ''::text) AND (feature_id IS NOT NULL) AND (feature_id <> ''::bpchar)));
```
Comment (gergely-kurucz-konghq) [inline tools/migrate/migrations/20260818090123_rate_card_feature_reference_constraints.up.sql:4]: Nit: Not sure why `bpchar` is used here, seems confusing at first glance.

File: tools/migrate/rate_card_feature_reference_constraints_test.go [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+				key:        "key_only",
+				featureID:  nil,
+				featureKey: featureKey,
+			}))
+			require.Error(t, insertRateCardFeatureReferenceConstraintRow(t, db, rateCardFeatureReferenceConstraintRow{
+				table:      table,
+				namespace:  namespace,
+				key:        "empty_key",
+				featureID:  featureID,
+				featureKey: "",
```
Comment (gergely-kurucz-konghq) [inline tools/migrate/rate_card_feature_reference_constraints_test.go:202]: Nit: Why is the
```
				featureID:  "",
				featureKey: featureKey,
```
case not tested?

**Review by gergely-kurucz-konghq — DISMISSED** [review]

File: tools/migrate/migrations/20260818090123_rate_card_feature_reference_constraints.up.sql (hunk 1 above)
Comment (chrisgacsal) [reply tools/migrate/migrations/20260818090123_rate_card_feature_reference_constraints.up.sql:4]: `feature_id` is stored as `character(26)`, and `bpchar `is PostgreSQL’s internal name for that fixed-width, blank-padded type. Atlas adds the explicit cast when normalizing the Ent check `feature_id <> ''`.

File: tools/migrate/rate_card_feature_reference_constraints_test.go (hunk 2 above)
Comment (chrisgacsal) [reply tools/migrate/rate_card_feature_reference_constraints_test.go:202]: Fixed.

File: tools/migrate/rate_card_feature_reference_constraints_test.go [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+			}))
```
[the remark this answers is not in the document]
Comment (chrisgacsal) [reply tools/migrate/rate_card_feature_reference_constraints_test.go:203]: Fixed.

**Review by gergely-kurucz-konghq — APPROVED** [review]
